Revised Special & Global Functions layout.

philmoz opened the pull request:
Summary of changes:
- Convert the page to use the flex layout
- Move the function name into the button control for an active function so the entire row acts as a button
- Displays the active functions first with fixed height for each line
- Displays the inactive functions second as a grid so there is less to scroll through

![Screen Shot 2023-01-06 at 6 30 30 pm](https://user-images.githubusercontent.com/9474356/210953807-1de85ea8-94c7-45dd-9d68-f7797a73fe92.png)

![Screen Shot 2023-01-06 at 6 30 46 pm](https://user-images.githubusercontent.com/9474356/210953832-8a7ee9f4-8444-4232-bc66-b0cdd8f31863.png)

froqstar commented:
What if, instead of listing all the inactive special functions, we just have an "add" button at the bottom like we have with mixes?

philmoz commented:
You can select one of the active functions and then choose 'Copy' from the menu.
Then on an inactive one you can choose 'Paste' from the menu.

With just an 'Add' button it would be more difficult to manage the copy/paste.

Also I like to use certain function numbers for the same operation across different models. I find it easier to manage this way. With just an 'Add' button it would be harder to pick a specific function to edit.

Eldenroot commented:
Nice improvement

ParkerEde commented:
This implementation is certainly good for touch operation. However, it is a degradation for non-touch color displays. I have for example from SF2 to SF20 some functions and in the range 50 to 64. If I now want to create SF1, you have to scroll a lot (up or down) to get to the button SF1. This implementation should only be used for touch radios, if at all. I personally have the feeling that with this implementation a lot of clarity is lost.

Eldenroot commented:
OK, but I think it is a good start :) we ca discuss :)

pfeerick commented:
> However, it is a degradation for non-touch color displays. I have for example from SF2 to SF20 some functions and in the range 50 to 64. If I now want to create SF1, you have to scroll a lot (up or down) to get to the button SF1.

How is that different from the current behaviour, in which a similar complaint could be made, since it is a fixed table, and you have to scroll through all the rows regardless of if there are entries present or not, rather than just scroll through the rows with content? 

However, I do get that coming from fixed table view, how this can be seen as a loss of clarity, as sometimes you leave blank SFs so you can group SFs related together, etc. 

> You can select one of the active functions and then choose 'Copy' from the menu.
Then on an inactive one you can choose 'Paste' from the menu.

This is actually something I find lacking in the new Mixer layout... before, you could copy an existing mixer line into a specific new one... i.e. what if I want to assign something to channel 32 on a five-channel model? (yes, that is a thing - for some special mixes and working with some lua scripts). Before I could just copy, scroll to 32, and paste. Now I have to create 32, then copy and well, um... there's no paste anymore (only paste before and after)... ok... that's because paste only appears for *empty* mixer lines... oh dear.

philmoz commented:
Away from main system at the moment so I can't try things; but a couple of thoughts:
- long press of enter button in SF/GF window to scroll to the first inactive function
- long press to bring up menu with options like 'Select first inactive', 'Add new', ...

philmoz commented:
I have added a long press handler that will jump to the first free function (if pressed on an active function), or the first active function if pressed on a free function.

ParkerEde commented:
Yes, this is now much easier to use thanks to the long press handler. Really good. One small problem has now come to my attention.
Assumed I have 10 SF rows. If I "delete" the second table entry (not "delete row"), the cursor jumps to the second free SF button. I think it would be clearer if it jumps to the next existing entry. What do think about it?

philmoz commented:
@ParkerEde The code currently tries to re-select the entry number that was focussed before any operation.

For example suppose you have SF1, SF2 and SF4 as active functions and SF3 as an empty one.
If you delete SF2, then all the functions move up 1, so SF3 becomes SF2, SF4 becomes SF3, etc.

When SF2 is re-selected as the focussed function it is still any empty one so the screen jumps to show it.

If you clear an active function it will become empty; but remain focussed so the screen will again scroll to the empty function section.

ParkerEde commented:
@philmoz 
Just to make sure there is no misunderstanding here. I have only selected "Clear". The following lines do not move up and the cursor jumps away. Here I would find it clearer, it would simply jump to the next existing entry.

If I select "Delete" , where SF3 then becomes SF2, SF4 then SF3 etc., it already works correctly.

philmoz commented:
@ParkerEde I understand what you are saying - I was just trying to explain how the logic currently works. I need to think about it more to see if I can come up with an alternative.

philmoz commented:
After using this some more, and trying it with some non-technical club members, I've decided to revert the list to the original numerical sort order. Although the grouped order looked nicer it is not intuitive to use.

Where possible the empty functions are still grouped on a single line to reduce the vertical size.

I have also made the button press handling consistent by always having a popup context menu. As well as being consistent behaviour it means you can use the insert and delete actions on empty function to manage the list.

![Screen Shot 2023-01-11 at 8 32 56 am](https://user-images.githubusercontent.com/9474356/211669769-95baa1c3-1c8c-42d4-8e06-6e13193fcb0f.png)

![Screen Shot 2023-01-11 at 8 33 17 am](https://user-images.githubusercontent.com/9474356/211669826-70ec7486-c003-4b96-a610-d64bf4dc9056.png)

ParkerEde commented:
I like it better that way, too. The overview is also restored. "Clear" and "Delete" also works well again. Good work.

ParkerEde commented:
I've checked last state 66d9f545 
The cursor always jumps to the top after "Clear" or "Delete". With "Clear" the same line should remain marked and with "Delete" the next entry present in the list should be jumped to.

philmoz commented:
@ParkerEde Thanks for testing - should be fixed now.

JimB40 commented:
@philmoz Please unify lists visually changing to flex. 
Visually SF List is same like Outputs. 
Part of 2.9 goal is to unify UI/UX

In terms of indexing (if to show free numeric slots, how to design UX of adding & displaying)  we have to discuss

philmoz commented:
@JimB40 Do you mean like this? If I use the button style from the the model outputs page then functions that are active (i.e. their switch is on) show with the background as light blue (secondary3 colour) instead of yellow (active colour).

![Screen Shot 2023-01-12 at 5 00 15 pm](https://user-images.githubusercontent.com/9474356/211989805-1c653007-f193-4d42-93ef-87c9d06207a2.png)

JimB40 commented:
As agreed. Use Inpusts/Mixes styling for list and I'm making issue to fix ACTIVE color in this style.

ParkerEde commented:
checked a98e9b3
It looks good for me now.

JimB40 commented:
Btw @philmoz I have checked data variances and it can be hard to fit all of them in one line

Examples of longest
- Switch - !WWWW (telemetry name) + switch icon
- Func - "Set Main Screen" longest

Then parameters depoendent on Func type
Set
- Timer - "Timer 3"
- Value - "08:59:59

Adjust
- Global var  "GV5"
- Mode - "Inc/Decrement"
- Inc/Decrement - "+= 2048"

Play Track
- Value - 'WWWWWW' (6 characters) 

Set Main Screen
- Value - 6

Then for some of functions
Repeat - "60s"

For others
Enable

By the way does anybody know what is purpose on Enable switch???
This feature is implemented for half of functions types

pfeerick commented:
> By the way does anybody know what is purpose on Enable switch???
> This feature is implemented for half of functions types

Literally that - it enables or disables that SF/GF. The fact that it is not implemented for all is annoying... i.e. I'd like to have a Screenshot GF, but disable it so that it's configured but doesn't hog the switch until I want it... but guess what SF/GF doesn't have the enable toggle? 😆

JimB40 commented:
Ok I thought so. Just confirming.
Do you know @pfeerick why it's not implemented for some of functions like Screenshot you mentioned? :)
Overlooked during coding?

philmoz commented:
> I have checked data variances and it can be hard to fit all of them in one line

I guess that's why it uses two lines to display special/global functions and logical switches.
It would be almost impossible to fit on one line on a portrait orientation screen.

JimB40 commented:
portrait is easier with data overview as it has a lot vertical space.
I think we have to go with flex -auto increasing line height and moving data elements to the second line if there are more elements) as there are too many options to fix positions.

ParkerEde commented:
how would it be to be able to assign individual descriptive names for SF and LS and then display only this in the row?

pfeerick commented:
> how would it be to be able to assign individual descriptive names for SF and LS and then display only this in the row?

This is where #2085 would need to be considered... 

> Do you know @pfeerick why Enable is not implemented for some of functions like Screenshot you mentioned? :)

You'd have to ask JC ... it's only been like this since I can started using OTX back in 2014 😆

Eldenroot commented:
I think that the best way is to create a new ticket about missing switches. And after that unify the gui and add them where missing.

ParkerEde commented:
I noticed one more little thing. If I create a SF that is displayed in the bottom row, but not the first from the row, then this is now marked but outside the viewing area.
![Zwischenablage01](https://user-images.githubusercontent.com/15068446/212060543-f7149d96-c615-4e71-af35-25ab8c8ecc73.jpg)
![Zwischenablage02](https://user-images.githubusercontent.com/15068446/212060557-97c6adc7-c41d-4bf1-a28f-8cc2d75b28d4.jpg)
![Zwischenablage03](https://user-images.githubusercontent.com/15068446/212060575-d01eec6b-c0bd-4437-b0d5-79ffbb13da3e.jpg)
![Zwischenablage04](https://user-images.githubusercontent.com/15068446/212060585-34c51bb9-152b-418f-890d-fb195899dda0.jpg)

philmoz commented:
@ParkerEde The scroll position should be fixed. Thanks again for testing.

JimB40 commented:
Guys still wondering if we need those empty slots shown.
I don't have full ETX data-relation overview yet, but to my best knowledge SF or GF is not used as the source anywhere else in the ETX. If you know it is please point me.
If it is not used it means SFxx or GFxx ID is just for information purpose.

ParkerEde commented:
@philmoz 
I have tested 59394aa. Now the created record is visible in the screen. I tested the same with the logical switches (PR3019), there it also works. Thanks

philmoz commented:
I have converted the landscape layout to show the overview in a single line by displaying the function as an expression instead of discrete values in fixed locations. IMHO this makes the functions easier to understand, and also means that all of the values / parameters for the function are displayed.
Portrait layout still uses two lines.

I also had to change the 'Enabled' checkbox because the old one was hard wired to paint directly into the bitmap buffer and was difficult to reproduce.

![Screen Shot 2023-01-14 at 8 18 30 am](https://user-images.githubusercontent.com/9474356/212424438-182b1748-1936-4b54-a071-86ef7ef1f469.png)

![Screen Shot 2023-01-14 at 8 17 40 am](https://user-images.githubusercontent.com/9474356/212424460-49962cf1-7b17-4696-b38d-2656f77366bb.png)

ParkerEde commented:
yeah, looks very nice.

but at the moment the radiobuttons are considered as list elements. For example, if SF63 and SF64 have an activation button, and you are currently on SF64, the next scroll point in SF63 is the radiobutton itself and receives the marker.
Small suggestion still. At the moment an activated radio button is completely filled. How about leaving a small area between the border and the fill unfilled? Then I think it will be a bit clearer that it represents an active radio button.

philmoz commented:
It's not a radio button - it's a small text button with no text (and marked as not selectable so you can't scroll to it).
I will see if I can come up with something that looks a bit different.

ParkerEde commented:
yes, really. the marker jumps to the bottons that look slike radiobuttons ;-)

https://user-images.githubusercontent.com/15068446/212433501-dd191ab0-a1e1-4acc-bd84-c893152db8bb.mp4

philmoz commented:
New version stops the 'enabled' indicator being selectable with the scroll wheel and changes the look.
If anyone can suggest a better look for the 'enabled' indicator I'm happy to try out other ideas.

![Screen Shot 2023-01-14 at 10 32 12 am](https://user-images.githubusercontent.com/9474356/212437404-934be6c0-59a5-49bc-83c3-ddcd3c3e7027.png)

ParkerEde commented:
yes great. Looks good.

philmoz commented:
Updated to match PR #3019.

Removed all the empty function buttons and added a single '+' button to match the inputs and mixes pages.

The '+' button will pop up the view to select an empty function unless 'Copy' has been previously selected on another function. In that case it will show a menu to allow pasting to an empty function (or creating a new one).

![Screen Shot 2023-01-14 at 12 53 39 pm](https://user-images.githubusercontent.com/9474356/212445544-afb39b58-56d9-4909-baed-b44cf9d34eb0.png)

![Screen Shot 2023-01-14 at 12 55 06 pm](https://user-images.githubusercontent.com/9474356/212445548-70db5a50-d570-45ad-826c-d808d808f8b0.png)

![Screen Shot 2023-01-14 at 12 55 19 pm](https://user-images.githubusercontent.com/9474356/212445554-5951ae00-a18c-48b3-9ccd-14c39402529b.png)

JimB40 commented:
@philmoz  Now you have those magic numbers, but less scrolling without empty slots. And look like we have good UI scheme for lists.
This '+' button is little inconvenient when in mid of 40 items list, but I'd definitely would leave it like that in 2.9
and globally change later to option in context menu in 3.0

Eldenroot commented:
The "+" button should be visible all the time, sticky.

philmoz commented:
There are probably a lot of other ways to tweak this (and PR #3019); but I think it is now fairly consistent with the inputs and mixes pages.

Should these be reviewed and merged first, then work on making them all better?

JimB40 commented:
Making '+" sticky will decrease data overview.

As I said this way or addind item to the list is temporary for 2.9
In 3.0 once close/back button is separated & moved to the right, screen icon will serve as button to open context menu for all screen's genereal functions like adding.

For now I'd not spend more time on that as UI result will be mediocre anyway and we have more advanced UI stuff to do for 2.9 release.

pfeerick commented:
> The "+" button should be visible all the time, sticky.

That is also only of value for touch - for scroll wheel only/hard key navigation there is no simple way to get to the button... ;)

Eldenroot commented:
You are right... so now we have "create new" in the pop up menu?

pfeerick commented:
https://github.com/EdgeTX/edgetx/pull/2992#issuecomment-1382631592 Shows the latest screenshots for this PR... and you have "New" and "Paste" options.

Eldenroot commented:
Ok, thank you. Looks nice, much better ;)

philmoz commented:
The 'New' and 'Paste' popup menu is activated by the '+' button; but only if another line has been copied previously. If the clipboard is empty the '+' button goes straight to the 'New' flow (pops up the list of empty functions).

You still have to scroll to the '+' button first.

On my TX16S the scroll wheel wraps around from the top to the bottom so you can get to the '+' more quickly that way.

philmoz commented:
I have added a long press handler to quickly access the '+' button functions.

ParkerEde commented:
I really like that now.
Can the longpresshandler still be applied to mixes and inputs?

philmoz commented:
It should be easy enough to add to the inputs and mixers pages - if there is general agreement that it is a reasonable solution.

There may be better uses for the long press.

gagarinlg commented:
@philmoz maybe it would be good when you start discussing with @JimB40 about the UI stuff. There is, after the LVGL transformation in 2.9 a big UI revamp planned for 3.0.
The amount of work you currently put into the UI should go in the right direction, so that is not thrown away when the work on the 3.0 UI starts.

mha1 commented:
@philmoz Thanks for all the good work you do!

While working on Special Functions and migrating Logical Switches: could you please have a look at the different column padding between those two. I think it would be nice to have a Special Functions padding with more room for the 1st column like in Logical Switches. So retaining the Logical Switches padding for the migration and adjust Special Functions to the same padding. Possible?

![image](https://user-images.githubusercontent.com/5615068/213667651-65c3376b-0840-4096-bc1c-5f1e262e3dfb.png)

mha1 commented:
Thanks!

pfeerick commented:
Final state of this looks and works great on TX16S and NV14. Thanks for all the discussion and work on this! :)
